mailpoet/lib/Automation/Integrations/MailPoet/Triggers/UserRegistrationTrigger.php

```diff
@@ -40,84 +42,52 @@ public function getArgsSchema(): ObjectSchema {
     $this->wp->addAction('mailpoet_user_registered', [$this, 'handleSubscription']);
   }
 
   public function handleSubscription(SubscriberEntity $subscriber): void {
-    $segment = $this->getSegment($subscriber);
+    $segment = $this->getWpSegment($subscriber);
     $this->wp->doAction(Hooks::TRIGGER, $this, [
-      [
-        'key' => SegmentSubject::KEY,
-        'args' => [
-          'segment_id' => $segment->getId(),
-        ],
-      ],
-      [
-        'key' => SubscriberSubject::KEY,
-        'args' => [
-          'subscriber_id' => $subscriber->getId(),
-        ],
-      ],
+      new Subject(SegmentSubject::KEY, ['segment_id' => $segment->getId()]),
+      new Subject(SubscriberSubject::KEY, ['subscriber_id' => $subscriber->getId()]),
     ]);
   }
 
-  public function isTriggeredBy(array $args, Subject ...$subjects): bool {
-    $segment = null;
-    $subscriber = null;
-    foreach ($subjects as $subject) {
-      if ($subject instanceof SegmentSubject) {
-        $segment = $subject->getSegment();
-      }
-      if ($subject instanceof SubscriberSubject) {
-        $subscriber = $subject->getSubscriber();
-      }
-    }
-
-    if (!$segment || !$subscriber) {
+  public function isTriggeredBy(StepRunArgs $args): bool {
+    $segmentPayload = $args->getSinglePayloadByClass(SegmentPayload::class);
+    if ($segmentPayload->getType() !== SegmentEntity::TYPE_WP_USERS) {
       return false;
     }
 
-    if ($segment->getType() !== SegmentEntity::TYPE_WP_USERS) {
+    $subscriberPayload = $args->getSinglePayloadByClass(SubscriberPayload::class);
+    if (!$subscriberPayload->isWPUser()) {
       return false;
     }
-    if (!isset($args['roles']) || !is_array($args['roles']) || !count($args['roles'])) {
-      return true;
-    }
 
-    if (!$subscriber->isWPUser()) {
-      return false;
-    }
-    $user = $this->wp->getUserBy('id', $subscriber->getWpUserId());
+    $user = $this->wp->getUserBy('id', $subscriberPayload->getWpUserId());
     if (!$user) {
       return false;
     }
 
-    foreach ($user->roles as $userRole) {
-      if (in_array($userRole, $args['roles'], true)) {
-        return true;
-      }
-    }
-    return false;
+    $triggerArgs = $args->getStep()->getArgs();
+    $roles = $triggerArgs['roles'] ?? [];
+    return !is_array($roles) || !$roles || count(array_intersect($user->roles, $roles)) > 0;
   }
 
-  private function getSegment(SubscriberEntity $subscriber): SegmentEntity {
-
-    $segments = $subscriber->getSubscriberSegments()->toArray();
-    if (!$segments) {
-      throw new InvalidStateException();
-    }
-    $segment = null;
-    foreach ($segments as $segment) {
-      $segment = $segment->getSegment();
-      if (!$segment || $segment->getType() !== SegmentEntity::TYPE_WP_USERS) {
-        continue;
-      }
-      break;
-    }
-    if (!$segment || $segment->getType() !== SegmentEntity::TYPE_WP_USERS) {
+  private function getWpSegment(SubscriberEntity $subscriber): SegmentEntity {
```

Comment:
@JanJakes Trying to create a new user results now in an error for me: 
![image](https://user-images.githubusercontent.com/6458412/193016423-6c1e5fa6-6768-4b41-bc78-3bb2a4ff5f5e.png)